src/main/java/org/openrewrite/staticanalysis/LambdaBlockToExpression.java

```diff
@@ -79,42 +81,66 @@ public J.MethodInvocation visitMethodInvocation(J.MethodInvocation method, Execu
         );
     }
 
+    private static boolean hasLambdaArgument(J.MethodInvocation method) {
+        boolean hasLambdaArgument = false;
+        for (Expression arg : method.getArguments()) {
+            if (arg instanceof J.Lambda) {
+                hasLambdaArgument = true;
+                break;
+            }
+        }
+        return hasLambdaArgument;
+    }
+
     // Check whether a method has overloading methods in the declaring class
-    private static boolean hasMethodOverloading(J.MethodInvocation method) {
+    static boolean hasAmbiguousMethodOverloading(MethodCall method) {
         JavaType.Method methodType = method.getMethodType();
-        return methodType != null && hasMethodOverloading(methodType);
+        return methodType != null
+                && hasAmbiguousMethodOverloading(methodType, method.getArguments(), methodType.getName());
     }
 
     // TODO this is actually more complex in the presence of generics and inheritance
-    static boolean hasMethodOverloading(JavaType.Method methodType) {
-        String methodName = methodType.getName();
-        return Optional.of(methodType)
+    static boolean hasAmbiguousMethodOverloading(JavaType.Method methodType, List<Expression> arguments, String methodName) {
+        int numberOfArguments = arguments.size();
+
+        //all methods of the given type
+        List<JavaType.Method> methodsOfType = Optional.of(methodType)
                 .map(JavaType.Method::getDeclaringType)
                 .filter(JavaType.Class.class::isInstance)
                 .map(JavaType.Class.class::cast)
                 .map(JavaType.Class::getMethods)
-                .map(methods -> {
-                    int overloadingCount = 0;
-                    for (JavaType.Method dm : methods) {
-                        if (dm.getName().equals(methodName)) {
-                            if (++overloadingCount > 1) {
-                                return true;
-                            }
-                        }
-                    }
-                    return false;
-                })
-                .orElse(false);
-    }
+                .orElse(Collections.emptyList());
```

Comment:
Rather than collecting all methods into a list, it would be more efficient to loop over them.

I am also saying this, because we should actually also be checking methods in super types and even default methods on implemented interfaces. But this can quickly get difficult, as I am not convinced that we will be reliable able to tell if a method is an override of some other method (and thus not an overload) when generics are involved. But once we cross that road, a recursive method which returns as soon as it finds a conflicting overload will be more efficient than first gathering all methods up front.

Reply:
I won't find time to finish this and have unassigned me. I trust your judgement that not collecting makes a enough of a difference to justify the loss in readability. I found the previous version very hard to get into (at first I thought it `map`s over the list elements, but it is always over the `Optional`): https://github.com/openrewrite/rewrite-static-analysis/blob/e57dac0991cf3e57edf99826d975c079d0971baf/src/main/java/org/openrewrite/staticanalysis/LambdaBlockToExpression.java#L80-L97 and so I hope the assignment to `MethodsOfType` can stay: https://github.com/openrewrite/rewrite-static-analysis/blob/758e7cfbec5e306f2e0b9487d9241ae4fe502f50/src/main/java/org/openrewrite/staticanalysis/LambdaBlockToExpression.java#L113-L118 Since the list is already there (from `getMethods`) and `Collections.emptyList()` is a constant, I think it makes no difference in performance while neatly separating the optional  stream from the loop over the elements.

Reply:
Yes, you are right, there is no additional list being created here. I was already thinking ahead for the use case with inherited method declarations. But let's forget about that for the moment then (a bug will probably get reported at some point).